lib/torrent.js

```diff
@@ -638,24 +704,110 @@ export default class Torrent extends EventEmitter {
     })
   }
 
-  _verifyPieces (cb) {
-    parallelLimit(this.pieces.map((piece, index) => cb => {
+  /**
+   * Callback called after a piece is verified
+   * @callback verifyPieceCallback
+   * @param {Error=} error
+   * @param {boolean=} isValid
+   */
+  /**
+   * Verify a single piece using hashing
+   * @param index
+   * @param {verifyPieceCallback} cb
+   * @private
+   */
+  _verifyPiece (index, cb) {
+    if (this.destroyed) return cb(new Error('torrent is destroyed'))
+
+    const getOpts = {}
+    // Specify length for the last piece in case it is zero-padded
+    if (index === this.pieces.length - 1) {
+      getOpts.length = this.lastPieceLength
+    }
+    this.store.get(index, getOpts, async (err, buf) => {
       if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-      const getOpts = {}
-      // Specify length for the last piece in case it is zero-padded
-      if (index === this.pieces.length - 1) {
-        getOpts.length = this.lastPieceLength
-      }
-      this.store.get(index, getOpts, async (err, buf) => {
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+      if (err) return queueMicrotask(() => cb(null, false)) // ignore error
 
-        if (err) return queueMicrotask(() => cb(null)) // ignore error
+      const hex = await hash(buf, 'hex')
+      if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-        const hex = await hash(buf, 'hex')
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+      cb(null, hex === this._hashes[index])
+    })
+  }
+
+  /**
+   * Verify pieces using bitfield, in case of in-congruences it will re-verify the file using hashing
+   * @param {callbackWithError} cb
+   * @private
+   */
+  _verifyPiecesUsingBitfield (cb) {
+    const piecesToCheck = new Set()
+    const piecesToFilesMap = new Map()
+    this.files.forEach(file => {
+      const previouslyVerifiedFilePieces = []
+      for (let i = file._startPiece; i <= file._endPiece; ++i) {
+        if (this.bitfield.get(i)) {
+          previouslyVerifiedFilePieces.push(i)
+          if (!piecesToFilesMap.has(i)) {
+            piecesToFilesMap.set(i, [])
+          }
+          piecesToFilesMap.get(i).push(file)
+          this.emit('verified', i)
+        } else {
+          this._markUnverified(i)
+          this._debug('piece invalid %s', i)
+        }
+      }
+      const nonOverlappingPiece = previouslyVerifiedFilePieces.find(piece => piecesToFilesMap.get(piece).length === 1)
+      if (nonOverlappingPiece) {
+        piecesToCheck.add(nonOverlappingPiece)
+      } else if (previouslyVerifiedFilePieces.length >= 1) {
+        piecesToCheck.add(previouslyVerifiedFilePieces[0])
+      }
+    })
+    const filesToCheck = new Set()
+    parallelLimit([...piecesToCheck].map((index) => cb => {
```

Comment:
mmm sucks u cannt use this._verifyPiecesUsingHash here

Reply:
I would love to, but the logic inside is different, `_verifyPiecesUsingHash` just sets the piece as verified or not, but the logic in line 770 is to check if the piece that the bitflied says it's verified is truly verified

Reply:
yep, i'm just making this for myself while thinking it over

Reply:
To re-use it there's the option to have `_verifyPiecesUsingHash` return in the callback a boolean array with the verification status

```js
 this._verifyPiecesUsingHash([...piecesToCheck], (err, piecesVerified) => {
      if (err) {
        return cb(err)
      }
      piecesVerified.forEach((isVerified, index) => {
        if (!isVerified) {
          const filesOnPiece = piecesToFilesMap.get(index)
          filesOnPiece.forEach(file => {
            filesToCheck.add(file)
          })
        }
      })
      // Fifth step, if we need to recheck files we are going to recheck them
      if (filesToCheck.length) {
        const piecesToRecheck = []
        filesToCheck.forEach((file) => {
          for (let i = file._startPiece; i <= file._endPiece; ++i) {
            if (piecesToRecheck.indexOf(i) === -1) {
              piecesToRecheck.push(i)
            }
          }
        })
        return this._verifyPiecesUsingHash(piecesToRecheck, cb)
      }
      cb(null)
    })
```

Reply:
Rewrote it a different way, we don't need the data in the callback cause we already have the info in the mutated bitfield